File: modules/portfolio-tracker/src/utils/memory.py
import json
import os
import threading
from pathlib import Path


class MemoryStore:
    def __init__(self, mappings_path: str = "data/mappings.json"):
        self._path = Path(mappings_path)
        self._lock = threading.Lock()
        self._data: dict[str, dict[str, str]] = {
            "securities": {},
            "accounts": {},
            "categories": {},
            "brokers": {},
        }
        self._load()

    def _load(self):
        if not self._path.exists():
            return
        try:
            with open(self._path) as f:
                loaded = json.load(f)
                for key in self._data:
                    if key in loaded:
                        self._data[key] = loaded[key]
        except (json.JSONDecodeError, OSError):
            pass

    def _save(self):
        self._path.parent.mkdir(parents=True, exist_ok=True)
        with open(self._path, "w") as f:
            json.dump(self._data, f, indent=2, sort_keys=True)

    def learn(self, mapping_type: str, key: str, value: str):
        key_lower = key.strip().lower()
        with self._lock:
            if mapping_type in self._data:
                self._data[mapping_type][key_lower] = value.strip()
                self._save()

    def recall(self, mapping_type: str, key: str) -> str | None:
        key_lower = key.strip().lower()
        return self._data.get(mapping_type, {}).get(key_lower)

    def recall_all(self, mapping_type: str) -> dict[str, str]:
        return dict(sorted(self._data.get(mapping_type, {}).items()))

    def forget(self, mapping_type: str, key: str):
        key_lower = key.strip().lower()
        with self._lock:
            if mapping_type in self._data and key_lower in self._data[mapping_type]:
                del self._data[mapping_type][key_lower]
                self._save()

File: modules/portfolio-tracker/src/memory.js
/**
 * JSON-based persistent mapping store for learned associations.
 * Ported 1:1 from src/utils/memory.py
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs';
import { dirname } from 'path';

export class MemoryStore {
  constructor(mappingsPath = 'data/mappings.json') {
    this._path = mappingsPath;
    this._data = {
      securities: {},
      accounts: {},
      categories: {},
      brokers: {},
    };
    this._load();
  }

  _load() {
    if (!existsSync(this._path)) return;
    try {
      const loaded = JSON.parse(readFileSync(this._path, 'utf8'));
      for (const key of Object.keys(this._data)) {
        if (loaded[key]) {
          this._data[key] = loaded[key];
        }
      }
    } catch {
      // Ignore parse errors, keep defaults
    }
  }

  _save() {
    mkdirSync(dirname(this._path), { recursive: true });
    writeFileSync(this._path, JSON.stringify(this._data, null, 2), 'utf8');
  }

  /**
   * Persistently learn an association.
   */
  learn(mappingType, key, value) {
    const keyLower = key.trim().toLowerCase();
    if (this._data[mappingType]) {
      this._data[mappingType][keyLower] = value.trim();
      this._save();
    }
  }

  /**
   * Recall a learned association.
   * @returns {string|null}
   */
  recall(mappingType, key) {
    const keyLower = key.trim().toLowerCase();
    return this._data[mappingType]?.[keyLower] ?? null;
  }

  /**
   * Recall all mappings of a type.
   * @returns {object}
   */
  recallAll(mappingType) {
    const entries = Object.entries(this._data[mappingType] || {});
    entries.sort(([a], [b]) => a.localeCompare(b));
    return Object.fromEntries(entries);
  }

  /**
   * Forget a learned association.
   */
  forget(mappingType, key) {
    const keyLower = key.trim().toLowerCase();
    if (this._data[mappingType] && keyLower in this._data[mappingType]) {
      delete this._data[mappingType][keyLower];
      this._save();
    }
  }
}
